Name-related enhancements, Title Screen translation fixes

xoascf opened the pull request:
- [x] Use a better way to change the name
- [x] Use correct characters for Title Screen
- [x] Add enhancement to change default player name

Out of scope for this PR:
~- [ ] Maybe add the name change to Crowd Control~
~- [ ] Use a region-agnostic name format in save files~

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/514999946.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/514999948.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/514999949.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/514999950.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/514999952.zip)
<!--- section:artifacts:end -->

xoascf committed:
Fix Title Screen text character set

xoascf committed:
Improve name change method

xoascf committed:
Use macro for max name characters

xoascf committed:
Add Default Player Name enhancement

xoascf committed:
Use extended character set

jbodner09 commented on soh/soh/Enhancements/presets.h:
```diff
@@ -347,10 +350,12 @@ const std::vector<PresetEntry> enhancedPresetEntries = {
     // Text Speed (1 to 5)
     PRESET_ENTRY_S32("gTextSpeed", 5),
     // King Zora Speed (1 to 5)
-    PRESET_ENTRY_S32("gMweepSpeed", 2),
```
These preset changes to king zora speed and link's pause menu animation aren't part of your intended change; please undo them in your next push.

jbodner09 commented on soh/soh/Enhancements/debugger/debugSaveEditor.cpp:
```diff
@@ -283,45 +283,71 @@ void DrawGroupWithBorder(T&& drawFunc) {
-        ImGui::InputScalar(nameID.c_str(), ImGuiDataType_U8, &gSaveContext.playerName[i], &one, NULL);
+    char* name = new char[NAME_MAX];
+    strcpy(name, GetPlayerName().c_str());
+    if (ImGui::InputText("Name", name, NAME_MAX + 1)) {
```
If you're intending for NAME_MAX + 1 here to leave room for the string's null terminator, then this isn't doing what you want.  Instead, you're letting people input a name that's one character larger than the buffer you've allocated!

Instead, just use NAME_MAX in `ImGui::InputText`, and then when you declare the name variable two lines above it, use `calloc(NAME_MAX + 1, sizeof(char));`  That will zero out the buffer so that the null terminator is always wherever it needs to be.

xoascf replied:
I'm not sure, but it seems to be the intended design, that +1 is for the null terminator, added automatically:
https://discourse.dearimgui.org/t/prevent-from-entering-more-than-x-characters-in-inputtext/102/2

jbodner09 replied:
You'll still definitely need the underlying `name` buffer to be NAME_MAX + 1 though

xoascf replied:
Well, not necessarily, since it will not be copied or used in the character array for the player's name.

xoascf replied:
If you use a buffer of `NAME_MAX` (8) for `ImGui::InputText`, you can only type 7 characters maximum (with the version that SoH currently has), that's another reason why this is used.

jbodner09 replied:
You're using `strcpy` on that buffer, which requires it to be null terminated AND have enough space for said terminator to be copied as well:  https://cplusplus.com/reference/cstring/strcpy/.

xoascf replied:
~Perfect, then I'll use `memcpy` as the game does.~
Yeah, it seems that ImGui doesn't like `memcpy`, (but it was because of the missing terminator), anyway, I ended up sticking with `strcpy` but with your suggestion.

Man, I really hate not reading the documentation before using a library function. Thanks.
~Feel free to holler(? at me if I get anything wrong again 😬~

jbodner09 commented on soh/soh/UIWidgets.cpp:
```diff
@@ -663,4 +663,52 @@ namespace UIWidgets {
+            DisableComponentSwitch(disabledTooltipText, alpha);
+        }
+
+        const auto bText = new char[bufferSize];
```
Same issue here as in the save editor; you need `calloc(bufferSize + 1, sizeof(char));`

jbodner09 commented on soh/soh/UIWidgets.cpp:
```diff
@@ -663,4 +663,52 @@ namespace UIWidgets {
+        const auto bText = new char[bufferSize];
+        strcpy(bText, val);
+
+        if (ImGui::InputText(id, bText, bufferSize + 1)) {
```
Also same issue as in the save editor; just use bufferSize, not bufferSize + 1

jbodner09 commented on soh/soh/UIWidgets.cpp:
```diff
@@ -663,4 +663,52 @@ namespace UIWidgets {
+            ImGui::PushStyleVar(ImGuiStyleVar_Alpha, alpha);
+        }
+
+        ImGui::Text(text, val);
```
This isn't correct at all.  Since `text` isn't a format string the way you're using it, you don't need the second `var` argument like the Enhancement Sliders, for example, do.  just do `ImGui::Text(text)`

jbodner09 commented on soh/src/overlays/gamestates/ovl_file_choose/file_choose.h:
```diff
@@ -208,11 +208,6 @@ void FileChoose_DrawOptions(GameState* thisx);
-u8 HandleMouseCursor(FileChooseContext* this, Input* input, int minx, int miny, int maxx, int maxy);
-Vec2f HandleMouseCursorSplit(FileChooseContext* this, Input* input, int minx, int miny, int maxx, int maxy, int countx,
-                             int county);
-
```
This also looks unrelated to your intended change and shouldn't be removed

jbodner09 commented on soh/src/overlays/gamestates/ovl_file_choose/z_file_choose.c:
```diff
@@ -4,9 +4,7 @@
 #include "textures/parameter_static/parameter_static.h"
-#include <textures/icon_item_static/icon_item_static.h>
 #include "soh/frame_interpolation.h"
-#include <GameVersions.h>
```
These header removals look unintentional as well and should be undone

jbodner09 commented on soh/src/overlays/gamestates/ovl_file_choose/z_file_choose.c:
```diff
@@ -497,10 +495,10 @@ void FileChoose_UpdateMainMenu(GameState* thisx) {
                     this->charIndex = 0;
                     this->charBgAlpha = 0;
-                    this->newFileNameCharCount = 0;
+                    this->newFileNameCharCount = strlen(defName) >= NAME_MAX ? NAME_MAX - 1 : strlen(defName) % NAME_MAX;
```
This ternary statement isn't necessary.  Since we've already guaranteed that players can only input at most NAME_MAX characters into the cvar, the cvar will be properly null terminated, and SetPlayerName loops through exactly NAME_MAX characters, we can just set newFileNameCharCount = strlen(defName).

xoascf replied:
If I do that, when I put a name of 8 characters, the cursor goes to the beginning, and not on the last character.

jbodner09 replied:
If you're trying to fix ImGui behavior for setting the cursor, it would be more proper to use ImGui code to do that instead of messing with the underlying game data since that character count is used elsewhere, and there could be unintended consequences for having it not exactly match the actual number of characters in the name.

xoascf replied:
No, typing a name with 8 characters in the control has the same behavior as putting a variable and using it there with `strlen()`. But I don't need `% NAME_MAX` anymore, so I'll remove it.

xoascf replied:
By the way, I'm referring to the File Select cursor that highlights the space after the last character typed, so that way you can delete it as it is in games that use "Link" as the default name.

jbodner09 commented on soh/src/overlays/gamestates/ovl_file_choose/z_file_choose.c:
```diff
@@ -707,10 +705,10 @@ void FileChoose_UpdateQuestMenu(GameState* thisx) {
         this->charIndex = 0;
         this->charBgAlpha = 0;
-        this->newFileNameCharCount = 0;
+        this->newFileNameCharCount = strlen(defName) >= NAME_MAX ? NAME_MAX - 1 : strlen(defName) % NAME_MAX;
```
Same here, just strlen should be fine.

briaguya0 commented:
Since this is adding new functionality (default names) it should be pointed to dev instead of bradley

Also might we worth renaming this PR "name related enhancements/fixes" or something like that to better reflect the scope

xoascf committed:
Update debugSaveEditor.cpp

TODO: Use the new extended character set for names throughout the game and update the format of save files to avoid converting most characters.

jbodner09 commented:
One last thing, I think when you did the branch off of develop-bradley instead of just develop, you included a few changes from other PRs that you don't need in this one.  Did you intentionally change `OTRGlobals.cpp`, `z_kanfont.c`, or `z_en_mag.c`?  If not, then please revert any changes to those files, and I'll say you're good to go.

xoascf commented:
> One last thing, I think when you did the branch off of develop-bradley instead of just develop, you included a few changes from other PRs that you don't need in this one. Did you intentionally change `OTRGlobals.cpp`, `z_kanfont.c`, or `z_en_mag.c`? If not, then please revert any changes to those files, and I'll say you're good to go.

The changes were intentional, and were there before:

~`OTRGlobals.cpp`: Restored a function (original from #2072, removed in #2223) that will possibly be replaced in the future in LUS anyway.~ It is already changed, the restoration will be reversed.
`z_kanfont.c`: All the necessary characters are used both for the name and for the PRESS START and NO CONTROLLER texts in other languages to be corrected; both are corrected simultaneously.
`z_en_mag.c`: The texts mentioned above are corrected, and `strlen()` is used instead of an unnecessary function.

jbodner09 commented:
Okay, so what's really going on is that this PR is doing multiple things.  The bulk of the changes are for adding a "default" name that's editable later on, but you're ALSO fixing an issue with the title screen and re-adding some code that was incorrectly removed.  In that case, please rename this PR so the title reflects that, and also make sure everything you did is described in more detail in the description (like, WHY did you re-add that removed code?)

xoascf committed:
Fix compilation issues

briaguya0 commented:
i'm going to close this one. the main reason is that it is doing multiple things, and people don't agree that everything included in this PR should be merged. if you spit it off into multiple PRs (doing one thing each) there's a better chance of some of it getting merged.

aMannus commented:
I'd also like to add that it'd be nice to have a bit more clarification on what problems it's trying to solve. `Add enhancement to change default player name` Why is this needed for example? It'd be appreciated if this was done for future PRs.
